Take any csv/text file of 2+ GB of your choice. --- (You can do this assignment on Google colab)

Read the file ( Present approach of reading the file )

Try different methods of file reading eg: Dask, Modin, Ray, pandas and present your findings in term of computational efficiency

Perform basic validation on data columns : eg: remove special character , white spaces from the col name

As you already know the schema hence create a YAML file and write the column name in YAML file. --define separator of read and write file, column name in YAML

Validate number of columns and column name of ingested file with YAML.

Write the file in pipe separated text file (|) in gz format.

Create a summary of the file:

Total number of rows,

total number of columns

file size

In [5]:
import pandas as pd
import time
import modin.pandas as pd
import ray
from dask import dataframe as dd
import re

import yaml

In [6]:
datapath = 'F:/data_glacier_internship/week6/ds/horror_movies.csv'
df = pd.read_csv(datapath)

To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init()

2023-02-12 01:55:38,061	INFO worker.py:1538 -- Started a local Ray instance.
2023-02-12 01:55:43,414	WARNING __init__.py:183 -- DeprecationWarning: `ray.worker.global_worker` is a private attribute and access will be removed in a future Ray version.


In [7]:
df.head()

Unnamed: 0      id      original_title               title  \
0           1  760161  Orphan: First Kill  Orphan: First Kill   
1           2  760741               Beast               Beast   
2           3  882598               Smile               Smile   
3           4  756999     The Black Phone     The Black Phone   
4           5  772450          Presencias           Presences   

  original_language                                           overview  \
0                en  After escaping from an Estonian psychiatric fa...   
1                en  A recently widowed man and his two teenage dau...   
2                en  After witnessing a bizarre, traumatic incident...   
3                en  Finney Blake, a shy but clever 13-year-old boy...   
4                es  A man who loses his wife and goes to seclude h...   

                                            tagline release_date  \
0  There's always been something wrong with Esther.   2022-07-27   
1                                 Fight for family.   2022-08-11   
2                   Once you see it, it’s too late.   2022-09-23   
3                          Never talk to strangers.   2022-06-22   
4                                               NaN   2022-09-07   

                        poster_path  popularity  ...  vote_average  \
0  /pHkKbIRoCe7zIFvqan9LFSaQAde.jpg    5088.584  ...           6.9   
1  /xIGr7UHsKf0URWmyyd5qFMAq4d8.jpg    2172.338  ...           7.1   
2  /hiaeZKzwsk4y4atFhmncO5KRxeT.jpg    1863.628  ...           6.8   
3  /lr11mCT85T1JanlgjMuhs9nMht4.jpg    1071.398  ...           7.9   
4  /dgDT3uol3mdvwEg0jt1ble3l9hw.jpg    1020.995  ...           7.0   

       budget      revenue  runtime    status  adult  \
0         0.0    9572765.0       99  Released  False   
1         0.0   56000000.0       93  Released  False   
2  17000000.0   45000000.0      115  Released  False   
3  18800000.0  161000000.0      103  Released  False   
4         0.0          0.0        0  Released  False   

                      backdrop_path                genre_names collection  \
0  /5GA3vV1aWWHTSDO5eno8V5zDo8r.jpg           Horror, Thriller   760193.0   
1  /2k9tBql5GYH328Krj66tDT9LtFZ.jpg   Adventure, Drama, Horror        NaN   
2  /mVNPfpydornVe4H4UCIk7WevWjf.jpg  Horror, Mystery, Thriller        NaN   
3  /AfvIjhDu9p64jKcmohS4hsPG95Q.jpg           Horror, Thriller        NaN   
4  /ojfzhdwRemcDt1I6pao6vVLw9AA.jpg                     Horror        NaN   

     collection_name  
0  Orphan Collection  
1                NaN  
2                NaN  
3                NaN  
4                NaN  

[5 rows x 21 columns]

## Reading data with Dask

In [22]:

start_time = time.perf_counter()
df = dd.read_csv('F:/data_glacier_internship/week6/ds/horror_movies.csv')
end_time = time.perf_counter()
print(f"Time to load data using pandas: {end_time - start_time:0.4f} seconds")

Time to load data using pandas: 0.1318 seconds


## Reading data with Pandas

In [23]:
start_time = time.perf_counter()
df = dd.read_csv('F:/data_glacier_internship/week6/ds/horror_movies.csv',delimiter=',')

end_time = time.perf_counter()
print(f"Time to load data using pandas: {end_time - start_time:0.4f} seconds")

Time to load data using pandas: 0.0161 seconds


## Reading data with Modin and Ray

In [24]:
ray.shutdown()
ray.init()

csv_file = "F:/data_glacier_internship/week6/ds/horror_movies.csv"
start_time = time.time()
data_frame = pd.read_csv(csv_file)
end_time = time.time()

print(f"Reading csv with Modin and Ray took {end_time - start_time} seconds.")

2023-02-12 04:16:59,571	INFO worker.py:1538 -- Started a local Ray instance.


Reading csv with Modin and Ray took 4.038899183273315 seconds.


### Dask exhibits the shortest reading time of 0.1318 seconds, outperforming Pandas, and Modin and Ray.

## Cleaning the Dataset

In [ ]:
df = df.dropna(how='any')

In [ ]:
df = df.applymap(lambda x: re.sub(r'[^\w\s]', '', str(x)))

In [ ]:
df.columns = df.columns.str.strip()

In [ ]:
df.head()

## Converting CSV to YAML

In [ ]:
df_panda = pd.DataFrame(df)

In [ ]:
df_dict = df_panda.to_dict()

In [ ]:
with open('F:/data_glacier_internship/week6/horror_movies.yaml', mode='w') as file:
    yaml.dump(df_dict, file, indent=2)

## Printing YAML

In [3]:
with open("F:/data_glacier_internship/week6/horror_movies.yaml", "r") as file:
    data = yaml.safe_load(file)

In [4]:
print(data)

{0: {0: 'Unnamed: 0', 1: 'id', 2: 'original_title', 3: 'title', 4: 'original_language', 5: 'overview', 6: 'tagline', 7: 'release_date', 8: 'poster_path', 9: 'popularity', 10: 'vote_count', 11: 'vote_average', 12: 'budget', 13: 'revenue', 14: 'runtime', 15: 'status', 16: 'adult', 17: 'backdrop_path', 18: 'genre_names', 19: 'collection', 20: 'collection_name'}}


### Both CSV and YAML contain 21 columns. So, the conversion went correct.

## Write the file in pipe separated text file (|) in gz format.

In [18]:
import gzip
import csv

# Open the csv file in read-only binary mode
with open("F:/data_glacier_internship/week6/ds/horror_movies.csv", 'rb') as f_in:
    # Create a csv reader object
    reader = csv.reader(f_in.read().decode("utf-8").splitlines(), delimiter=',')

    # Open a gz file in write mode
    with gzip.open("F:/data_glacier_internship/week6/ds/horror_movies.csv.gz", 'wb') as f_out:
        # Write the rows to the gz file
        for row in reader:
            line = '|'.join(row) + '\n'
            f_out.write(line.encode("utf-8"))




In [19]:
import os

# Path to the directory
directory = "F:/data_glacier_internship/week6/ds/"

# Get a list of all entries in the directory
entries = os.listdir(directory)

# Filter the list to only include files with .gz extension
gz_files = [entry for entry in entries if entry.endswith('.gz')]

# Print the number of gz files in the directory
print(f"Number of gz files in the directory: {len(gz_files)}")

Number of gz files in the directory: 1


## Summary

In [21]:
#size of the file:
os.path.getsize('F:/data_glacier_internship/week6/ds/horror_movies.csv.gz')

6026539